File: tests/basic_tests.cpp
// Basic module tests for fastdist library.
#include <cassert>
#include <cmath>
#include <iostream>
#include <limits>

#include <fastdist/math/bernoulli.h>
#include <fastdist/math/binomial.h>
#include <fastdist/math/discrete_uniform.h>
#include <fastdist/math/exponential.h>
#include <fastdist/math/geometric.h>
#include <fastdist/math/normal.h>
#include <fastdist/math/poisson.h>
#include <fastdist/math/uniform.h>

int main() {
    std::cout << "Running fastdist basic tests...\n";

    const double tol = 1e-12;
    const double nan = std::numeric_limits<double>::quiet_NaN();

    // -------------------------
    // Normal distribution tests
    // -------------------------
    {
        double pdf0 = fastdist::math::normal_pdf_scalar(0.0, 0.0, 1.0);
        assert(std::abs(pdf0 - 0.3989422804014327) < tol);

        double cdf0 = fastdist::math::normal_cdf_scalar(0.0, 0.0, 1.0);
        assert(std::abs(cdf0 - 0.5) < tol);

        assert(fastdist::math::normal_mean(2.0) == 2.0);
        assert(fastdist::math::normal_variance(2.0) == 4.0);

        assert(fastdist::math::z_score(1.0, 0.0, 2.0) == 0.5);
    }

    // ------------------------------
    // Exponential distribution tests
    // ------------------------------
    {
        double pdf0 = fastdist::math::exponential_pdf_scalar(0.0, 2.0);
        assert(std::abs(pdf0 - 2.0) < tol);

        double cdf0 = fastdist::math::exponential_cdf_scalar(0.0, 2.0);
        assert(std::abs(cdf0 - 0.0) < tol);

        assert(std::abs(fastdist::math::exponential_mean(2.0) - 0.5) < tol);
        assert(std::abs(fastdist::math::exponential_variance(2.0) - 0.25) < tol);
    }

    // -------------------------
    // Poisson distribution tests
    // -------------------------
    {
        double pmf0 = fastdist::math::poisson_pmf_scalar(0.0, 3.0);
        assert(std::abs(pmf0 - std::exp(-3.0)) < tol);

        double pmf3 = fastdist::math::poisson_pmf_scalar(3.0, 3.0);
        assert(std::abs(pmf3 - 0.22404180765538775) < tol);

        double cdf0 = fastdist::math::poisson_cdf_scalar(0.0, 3.0);
        assert(std::abs(cdf0 - std::exp(-3.0)) < tol);

        assert(fastdist::math::poisson_mean(3.0) == 3.0);
        assert(fastdist::math::poisson_variance(3.0) == 3.0);
    }

    // -------------------------
    // Bernoulli distribution tests
    // -------------------------
    {
        assert(std::abs(fastdist::math::bernoulli_pmf_scalar(1, 0.3) - 0.3) < tol);
        assert(std::abs(fastdist::math::bernoulli_pmf_scalar(0, 0.3) - 0.7) < tol);
        assert(fastdist::math::bernoulli_pmf_scalar(2, 0.3) == 0.0);

        assert(std::abs(fastdist::math::bernoulli_cdf_scalar(0, 0.3) - 0.7) < tol);
        assert(std::abs(fastdist::math::bernoulli_cdf_scalar(1, 0.3) - 1.0) < tol);

        assert(std::abs(fastdist::math::bernoulli_mean(0.3) - 0.3) < tol);
        assert(std::abs(fastdist::math::bernoulli_variance(0.3) - 0.21) < tol);
        assert(std::abs(fastdist::math::bernoulli_stddev(0.3) - std::sqrt(0.21)) < tol);
    }

    // -------------------------
    // Binomial distribution tests
    // -------------------------
    {
        double pmf2 = fastdist::math::binomial_pmf_scalar(2, 3, 0.5);
        assert(std::abs(pmf2 - 0.375) < tol);

        double cdf1 = fastdist::math::binomial_cdf_scalar(1, 3, 0.5);
        assert(std::abs(cdf1 - 0.5) < tol);

        assert(std::abs(fastdist::math::binomial_mean(3, 0.5) - 1.5) < tol);
        assert(std::abs(fastdist::math::binomial_variance(3, 0.5) - 0.75) < tol);
        assert(std::abs(fastdist::math::binomial_stddev(3, 0.5) - std::sqrt(0.75)) < tol);
    }

    // -------------------------
    // Discrete Uniform distribution tests
    // -------------------------
    {
        double pmf3 = fastdist::math::discrete_uniform_pmf_scalar(3, 1, 6);
        assert(std::abs(pmf3 - 1.0 / 6.0) < tol);

        double cdf4 = fastdist::math::discrete_uniform_cdf_scalar(4, 1, 6);
        assert(std::abs(cdf4 - 4.0 / 6.0) < tol);

        assert(std::abs(fastdist::math::discrete_uniform_mean(1, 6) - 3.5) < tol);
        assert(std::abs(fastdist::math::discrete_uniform_variance(1, 6) - 35.0 / 12.0) < tol);
        assert(std::abs(fastdist::math::discrete_uniform_stddev(1, 6) - std::sqrt(35.0 / 12.0)) < tol);
    }

    // -------------------------
    // Continuous Uniform distribution tests
    // -------------------------
    {
        double pdf0 = fastdist::math::uniform_pdf_scalar(0.5, 0.0, 1.0);
        assert(std::abs(pdf0 - 1.0) < tol);

        double pdf_out = fastdist::math::uniform_pdf_scalar(1.5, 0.0, 1.0);
        assert(std::abs(pdf_out - 0.0) < tol);

        double cdf0 = fastdist::math::uniform_cdf_scalar(0.5, 0.0, 1.0);
        assert(std::abs(cdf0 - 0.5) < tol);

        double cdf_out = fastdist::math::uniform_cdf_scalar(1.5, 0.0, 1.0);
        assert(std::abs(cdf_out - 1.0) < tol);

        assert(std::abs(fastdist::math::uniform_mean(0.0, 1.0) - 0.5) < tol);
        assert(std::abs(fastdist::math::uniform_variance(0.0, 1.0) - 1.0 / 12.0) < tol);
        assert(std::abs(fastdist::math::uniform_stddev(0.0, 1.0) - std::sqrt(1.0 / 12.0)) < tol);
    }

    // -------------------------
    // Geometric distribution tests
    // -------------------------
    {
        double pmf1 = fastdist::math::geometric_pmf_scalar(1, 0.25); // k=1
        assert(std::abs(pmf1 - 0.25) < tol);

        double pmf3 = fastdist::math::geometric_pmf_scalar(3, 0.25);
        assert(std::abs(pmf3 - (0.25 * std::pow(0.75, 2))) < tol);

        double cdf2 = fastdist::math::geometric_cdf_scalar(2, 0.25);
        assert(std::abs(cdf2 - (0.25 + 0.25 * 0.75)) < tol);

        double mean = fastdist::math::geometric_mean(0.25);
        assert(std::abs(mean - 4.0) < tol); // 1/p

        double var = fastdist::math::geometric_variance(0.25);
        assert(std::abs(var - 12.0) < tol); // (1-p)/p^2

        double stddev = fastdist::math::geometric_stddev(0.25);
        assert(std::abs(stddev - std::sqrt(12.0)) < tol);
    }

    std::cout << "All basic tests passed.\n";
    return 0;
}

File: tests/basic_tests.py
import math

import fastdist


def main():
    print("Running fastdist basic tests...")

    tol = 1e-12
    nan = float("nan")

    # -------------------------
    # Normal distribution tests
    # -------------------------
    pdf0 = fastdist.normal_pdf_scalar(0.0, 0.0, 1.0)
    assert abs(pdf0 - 0.3989422804014327) < tol

    cdf0 = fastdist.normal_cdf_scalar(0.0, 0.0, 1.0)
    assert abs(cdf0 - 0.5) < tol

    assert fastdist.normal_mean(2.0) == 2.0
    assert fastdist.normal_variance(2.0) == 4.0
    assert fastdist.z_score(1.0, 0.0, 2.0) == 0.5

    # ------------------------------
    # Exponential distribution tests
    # ------------------------------
    pdf0 = fastdist.exponential_pdf_scalar(0.0, 2.0)
    assert abs(pdf0 - 2.0) < tol

    cdf0 = fastdist.exponential_cdf_scalar(0.0, 2.0)
    assert abs(cdf0 - 0.0) < tol

    assert abs(fastdist.exponential_mean(2.0) - 0.5) < tol
    assert abs(fastdist.exponential_variance(2.0) - 0.25) < tol

    # -------------------------
    # Poisson distribution tests
    # -------------------------
    pmf0 = fastdist.poisson_pmf_scalar(0.0, 3.0)
    assert abs(pmf0 - math.exp(-3.0)) < tol

    pmf3 = fastdist.poisson_pmf_scalar(3.0, 3.0)
    assert abs(pmf3 - 0.22404180765538775) < tol

    cdf0 = fastdist.poisson_cdf_scalar(0.0, 3.0)
    assert abs(cdf0 - math.exp(-3.0)) < tol

    assert fastdist.poisson_mean(3.0) == 3.0
    assert fastdist.poisson_variance(3.0) == 3.0

    # -------------------------
    # Bernoulli distribution tests
    # -------------------------
    assert abs(fastdist.bernoulli_pmf_scalar(1, 0.3) - 0.3) < tol
    assert abs(fastdist.bernoulli_pmf_scalar(0, 0.3) - 0.7) < tol
    assert fastdist.bernoulli_pmf_scalar(2, 0.3) == 0.0

    assert abs(fastdist.bernoulli_cdf_scalar(0, 0.3) - 0.7) < tol
    assert abs(fastdist.bernoulli_cdf_scalar(1, 0.3) - 1.0) < tol

    assert abs(fastdist.bernoulli_mean(0.3) - 0.3) < tol
    assert abs(fastdist.bernoulli_variance(0.3) - 0.21) < tol
    assert abs(
        fastdist.bernoulli_stddev(0.3) - math.sqrt(0.21)
    ) < tol

    # -------------------------
    # Binomial distribution tests
    # -------------------------
    pmf2 = fastdist.binomial_pmf_scalar(2, 3, 0.5)
    assert abs(pmf2 - 0.375) < tol

    cdf1 = fastdist.binomial_cdf_scalar(1, 3, 0.5)
    assert abs(cdf1 - 0.5) < tol

    assert abs(fastdist.binomial_mean(3, 0.5) - 1.5) < tol
    assert abs(fastdist.binomial_variance(3, 0.5) - 0.75) < tol
    assert abs(
        fastdist.binomial_stddev(3, 0.5) - math.sqrt(0.75)
    ) < tol

    # -------------------------
    # Discrete Uniform distribution tests
    # -------------------------
    pmf3 = fastdist.discrete_uniform_pmf_scalar(3, 1, 6)
    assert abs(pmf3 - 1.0 / 6.0) < tol

    cdf4 = fastdist.discrete_uniform_cdf_scalar(4, 1, 6)
    assert abs(cdf4 - 4.0 / 6.0) < tol

    assert abs(fastdist.discrete_uniform_mean(1, 6) - 3.5) < tol
    assert abs(
        fastdist.discrete_uniform_variance(1, 6)
        - 35.0 / 12.0
    ) < tol
    assert abs(
        fastdist.discrete_uniform_stddev(1, 6)
        - math.sqrt(35.0 / 12.0)
    ) < tol

    # -------------------------
    # Continuous Uniform distribution tests
    # -------------------------
    pdf0 = fastdist.uniform_pdf_scalar(0.5, 0.0, 1.0)
    assert abs(pdf0 - 1.0) < tol

    pdf_out = fastdist.uniform_pdf_scalar(1.5, 0.0, 1.0)
    assert abs(pdf_out - 0.0) < tol

    cdf0 = fastdist.uniform_cdf_scalar(0.5, 0.0, 1.0)
    assert abs(cdf0 - 0.5) < tol

    cdf_out = fastdist.uniform_cdf_scalar(1.5, 0.0, 1.0)
    assert abs(cdf_out - 1.0) < tol

    assert abs(fastdist.uniform_mean(0.0, 1.0) - 0.5) < tol
    assert abs(
        fastdist.uniform_variance(0.0, 1.0) - 1.0 / 12.0
    ) < tol
    assert abs(
        fastdist.uniform_stddev(0.0, 1.0)
        - math.sqrt(1.0 / 12.0)
    ) < tol

    # -------------------------
    # Geometric distribution tests
    # -------------------------
    pmf1 = fastdist.geometric_pmf_scalar(1, 0.25)
    assert abs(pmf1 - 0.25) < tol

    pmf3 = fastdist.geometric_pmf_scalar(3, 0.25)
    assert abs(pmf3 - (0.25 * math.pow(0.75, 2))) < tol

    cdf2 = fastdist.geometric_cdf_scalar(2, 0.25)
    assert abs(cdf2 - (0.25 + 0.25 * 0.75)) < tol

    mean = fastdist.geometric_mean(0.25)
    assert abs(mean - 4.0) < tol

    var = fastdist.geometric_variance(0.25)
    assert abs(var - 12.0) < tol

    stddev = fastdist.geometric_stddev(0.25)
    assert abs(stddev - math.sqrt(12.0)) < tol

    # -------------------------
    # Chebyshev inequality test
    # -------------------------
    bound = fastdist.chebyshev_bound(4.0, 2.0)
    assert abs(bound - 1.0) < tol

    bad = fastdist.chebyshev_bound(4.0, 0.0)
    assert math.isnan(bad)

    print("All basic tests passed.")


if __name__ == "__main__":
    main()
